aten/src/THC/THCBlas.cu

```diff
@@ -238,6 +238,31 @@ void adjustLdLevel3(char transa, char transb, int64_t m, int64_t n, int64_t k, i
 
 }
 
+// Check https://gist.github.com/umanwizard/2b2e2fc12485ef6dc1cdfb1421276dd9
+// to repro the bug. We don't know the exact conditions that trigger it,
+// but using Sgemm or Hgemm on Maxwell or Pascal seems to be a
+// necessary condition.
+static void checkCuda90Bug(int i_m, int i_n, int i_k)
+{
+#if CUDA_VERSION < 9200 && CUDA_VERSION >= 9000
+  static bool alreadyWarned = false;
+  if (alreadyWarned) {
+    return;
+  }
+  const int LIMIT = 1 << 21;
+  if (i_m > LIMIT || i_n > LIMIT || i_k > LIMIT) {
+    cudaDeviceProp* prop = at::cuda::getCurrentDeviceProperties();
+    if (prop->major == 5 || prop->major == 6) {
+      if (!alreadyWarned) {
+        fprintf(stderr, "Matrix multiplication for dimensions larger than 2^21 has known bugs on your combination of CUDA version and device type. "
```

Comment:
Please use `TORCH_WARN`.

Reply:
Definitely agree with `TORCH_WARN`.

On a related note, do we have methods of suppressing warnings in C++ via `TORCH_WARN`?  I generally don't think doing things like only warning once is correct -- python already has a bunch of functionality around controlling repeating warnings, so reinventing it with no configuration it isn't a good idea.  I'm not really sure how that applies to C++, though.

Reply:
Thanks for the pointer to `TORCH_WARN`. I was trying to find a warning macro, but I was looking for `THWarn` (which doesn't exist) :)

Reply:
@gchanan 

> python already has a bunch of functionality around controlling repeating warnings

Yeah, I agree this isn't a great way of doing it, but I'm not sure what such facilities Python has -- can you elaborate?

Anyway, I went with the pattern of warning once based on a flag because I saw that pattern elsewhere in TH.

Reply:
See filtering here: https://docs.python.org/3/library/warnings.html.

Reply:
`TORCH_WARN` is nice because it actually triggers a native Python warning, so at least our main frontend users get a pleasant experience. I'm pretty sure that unless you enable Python this decays to `printf` in C++. It's definitely now nice and we should improve this, but I guess it's not a priority 🤷🏻‍♂️